flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultKeyedStateStore.java

```diff
@@ -35,21 +35,50 @@
 import org.apache.flink.api.common.state.ValueStateDescriptor;
 import org.apache.flink.util.Preconditions;
 
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
 import static java.util.Objects.requireNonNull;
+import static org.apache.flink.util.Preconditions.checkState;
 
 /**
  * Default implementation of KeyedStateStore that currently forwards state registration to a {@link
  * RuntimeContext}.
  */
 public class DefaultKeyedStateStore implements KeyedStateStore {
 
-    protected final KeyedStateBackend<?> keyedStateBackend;
+    @Nullable protected final KeyedStateBackend<?> keyedStateBackend;
+
+    @Nullable protected final AsyncKeyedStateBackend<?> asyncKeyedStateBackend;
     protected final SerializerFactory serializerFactory;
 
+    protected SupportKeyedStateApiSet supportKeyedStateApiSet;
+
     public DefaultKeyedStateStore(
             KeyedStateBackend<?> keyedStateBackend, SerializerFactory serializerFactory) {
-        this.keyedStateBackend = Preconditions.checkNotNull(keyedStateBackend);
+        this(keyedStateBackend, null, serializerFactory);
+    }
+
+    public DefaultKeyedStateStore(
+            AsyncKeyedStateBackend<?> asyncKeyedStateBackend, SerializerFactory serializerFactory) {
+        this(null, asyncKeyedStateBackend, serializerFactory);
+    }
+
+    public DefaultKeyedStateStore(
+            @Nullable KeyedStateBackend<?> keyedStateBackend,
+            @Nullable AsyncKeyedStateBackend<?> asyncKeyedStateBackend,
+            SerializerFactory serializerFactory) {
+        this.keyedStateBackend = keyedStateBackend;
+        this.asyncKeyedStateBackend = asyncKeyedStateBackend;
         this.serializerFactory = Preconditions.checkNotNull(serializerFactory);
+        if (keyedStateBackend != null) {
+            // By default, we support state v1
```

Comment:
Can STATE_V1 and STATE_V2 coexist in one job?
Maybe we can add some comments about this in the state document later.

Reply:
No they can't coexist. We will introduce this in a complete document later.